### PR TITLE
feat(http prober): #1285 implement cache http response

### LINKED ISSUE #1285: Cache the responses of identical requests to prevent multiple requests within a short time
To save network bandwidth, Monika needs to cache the responses of identical requests.

### PR DESCRIPTION
**Pull request by syamsudotdev**
# Monika Pull Request (PR)

This PR resolves #1285 

## What feature/issue does this PR add

1. Added new file `src/components/probe/prober/http/response-cache.ts` : in-memory cache for HTTP response

## How did you implement / how did you fix it

1. Create in-memory cache with `@isaac/ttl-cache`
2. Changed `HTTPProber` to use cached response on first probing attempt
3. Added new CLI option `--ttl-cache` to set cache's time-to-live
4. Added new CLI option `--verbose-cache` to monitor cache use

## How to test

1. `npm run start -- --verbose-cache`
2. Observe messages for `Cache HIT` and `Cache MISS`

image

### CODE REVIEW

File: src/components/probe/prober/http/response-cache.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * And default time-to-live for each cache entries is 30s.
+ *
+ * About cache size 50 MB limit :
+ * Assuming a typical web page response is around 500 KB
+ * This cache can fit around 100 entries of web pages
+ * A typical response with empty body is around 300 bytes of headers
+ * That means, this cache can fit around 160K entries of empty body responses
+ */
+const DEFAULT_CACHE_LIMIT = 50_000_000 // 50 MB in bytes
+const DEFAULT_TIME_TO_LIVE = 30_000 // 30s in ms
```
Comment (dennypradipta) [inline src/components/probe/prober/http/response-cache.ts:41]: Should the cache limit and TTL configurable through flags?

Comment (dennypradipta) [reply src/components/probe/prober/http/response-cache.ts:41]: I am thinking in the real case scenario where probing may be 5 minutes per probe, should the TTL be longer?

**Comment by haricnugraha** [comment]
Wouldn't it be easier if we use the library with tested code such as https://www.npmjs.com/package/@isaacs/ttlcache?

File: src/components/probe/prober/http/response-cache.ts (hunk 1 above)
Comment (syamsudotdev) [reply src/components/probe/prober/http/response-cache.ts]: updated

**Comment by syamsudotdev** [comment]
> Wouldn't it be easier if we use the library with tested code such as https://www.npmjs.com/package/@isaacs/ttlcache?

@haricnugraha Updated cache backend with said package

**Review by sapiderman — APPROVED** [review]
looks good.

**Comment by syamsudotdev** [comment]
> btw: Do we have memory consumption for say a cache of 100 probe reponses? 1000?

@sapiderman A typical web page like youtube and sentry.io is around 500 kilobytes for their html text.

```
curl -L https://sentry.io > test
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  557k    0  557k    0     0   363k      0 --:--:--  0:00:01 --:--:--  711k
```

For 100 probes of web pages, I'd say it would be around 50 megabytes.

File: docs/src/pages/guides/cli-options.md [hunk 2]
```diff
@@ -344,6 +355,14 @@ Like your app to be more chatty and honest revealing all its internal details? U
 monika --verbose
 ```
 
+## Verbose Cache
+
+Show (HTTP) cache hit / miss messages to log
+
+```bash
+monika --verbose-cache
```
Comment (raosan) [inline docs/src/pages/guides/cli-options.md:363]: weird, i got no cache hit/miss log
Screenshot 2024-06-11 at 13 39 03

Screenshot 2024-06-11 at 13 40 36

i use this `monika.yml`:
```
probes:
  - id: '1'
    name: GitHub
    description: Multiple
    requests:
      - url: https://github.com
        timeout: 7000
      - url: https://github.com/hyperjump
        timeout: 7000

db_limit:
  max_db_size: 1000000000
  deleted_data: 1
  cron_schedule: '*/1 * * * *'
```

Comment (raosan) [reply docs/src/pages/guides/cli-options.md:363]: turns out it is a chaining request.

works on single request
image


LGTM then...

**Review by raosan — APPROVED** [review]

File: docs/src/pages/guides/cli-options.md (hunk 2 above)
Comment (syamsudotdev) [reply docs/src/pages/guides/cli-options.md:365]: Added redundant documentation message for request chaining in `verbose-cache` for clearer explanation.

**Pull request merged**